File: apps/trading_dashboard/backend/app/schemas/execution.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel


class ExecutionStatus(BaseModel):
    log_dir: str
    health: dict[str, Any]
    lock: dict[str, Any]
    command: str | None = None
    account: dict[str, Any] | None = None
    latest_by_asset: list[dict[str, Any]]
    recent_events: list[dict[str, Any]]
    files: list[dict[str, Any]]


class ExecutionBotOption(BaseModel):
    id: str
    label: str
    log_dir: str
    resolved_log_dir: str
    config_path: str | None = None
    mode: str | None = None
    state: str
    pid: int | None = None
    process_running: bool | None = None
    last_heartbeat_at: str | None = None
    modified_at: str | None = None
    symbols: list[str]
    has_logs: bool
    is_default: bool


class ExecutionBotOptionList(BaseModel):
    options: list[ExecutionBotOption]


class ExecutionRecordList(BaseModel):
    log_dir: str
    records: list[dict[str, Any]]


class ExecutionFeatureSnapshot(BaseModel):
    log_dir: str
    asset: str
    mt5_symbol: str | None = None
    bar_time: str | None = None
    timeframe: str | None = None
    row_count: int
    columns: list[str]
    numeric_columns: list[str]
    feature_columns: list[str]
    market_columns: list[str]
    records: list[dict[str, Any]]


class MarketMakingSnapshot(BaseModel):
    run_dir: str
    asset: str
    row_count: int
    columns: list[str]
    numeric_columns: list[str]
    feature_columns: list[str]
    market_columns: list[str]
    records: list[dict[str, Any]]
    trades: list[dict[str, Any]]
    summary: dict[str, Any]


__all__ = [
    "ExecutionBotOption",
    "ExecutionBotOptionList",
    "ExecutionFeatureSnapshot",
    "ExecutionRecordList",
    "ExecutionStatus",
    "MarketMakingSnapshot",
]

File: apps/trading_dashboard/frontend/src/types/execution.ts
export type JsonRecord = Record<string, unknown>;

export interface ExecutionAssetSummary {
  asset: string;
  mt5_symbol?: string | null;
  bar_time?: string | null;
  logged_at?: string | null;
  close?: number | null;
  spread?: number | null;
  signal_side?: number | string | null;
  order_action?: string | null;
  order_status?: string | null;
  order_reason?: string | null;
  has_decision_trace: boolean;
}

export interface ExecutionStatus {
  log_dir: string;
  health: JsonRecord;
  lock: JsonRecord;
  command?: string | null;
  account?: JsonRecord | null;
  latest_by_asset: ExecutionAssetSummary[];
  recent_events: JsonRecord[];
  files: JsonRecord[];
}

export interface ExecutionBotOption {
  id: string;
  label: string;
  log_dir: string;
  resolved_log_dir: string;
  config_path?: string | null;
  mode?: string | null;
  state: string;
  pid?: number | null;
  process_running?: boolean | null;
  last_heartbeat_at?: string | null;
  modified_at?: string | null;
  symbols: string[];
  has_logs: boolean;
  is_default: boolean;
}

export interface ExecutionBotOptionList {
  options: ExecutionBotOption[];
}

export interface ExecutionRecordList {
  log_dir: string;
  records: JsonRecord[];
}

export interface ExecutionFeatureSnapshot {
  log_dir: string;
  asset: string;
  mt5_symbol?: string | null;
  bar_time?: string | null;
  timeframe?: string | null;
  row_count: number;
  columns: string[];
  numeric_columns: string[];
  feature_columns: string[];
  market_columns: string[];
  records: JsonRecord[];
}

export interface MarketMakingSnapshot {
  run_dir: string;
  asset: string;
  row_count: number;
  columns: string[];
  numeric_columns: string[];
  feature_columns: string[];
  market_columns: string[];
  records: JsonRecord[];
  trades: JsonRecord[];
  summary: JsonRecord;
}
